src/mesh/mesh.cxx

```diff
@@ -317,6 +318,89 @@ bool Mesh::hasBndryUpperY() {
   return answer;
 }
 
+int Mesh::localSize3D() {
+  if (localNumCells3D < 0) {
+    const int xs = firstX() ? xstart - 1 : xstart;
+    const int xe = lastX() ? xend + 2 : xend + 1;
+    const int nx = xe - xs;
+    const int ny = yend - ystart + 1;
+    const int nz = LocalNz;
+    localNumCells3D = nx * ny * nz;
+    for (RangeIterator it = iterateBndryLowerY(); !it.isDone(); it++) {
+      if (it.ind == xstart)
+        localNumCells3D += nz * ystart;
+      if (it.ind == xend)
+        localNumCells3D += nz * ystart;
+      localNumCells3D += nz * ystart;
+    }
+    for (RangeIterator it = iterateBndryUpperY(); !it.isDone(); it++) {
+      if (it.ind == xstart)
+        localNumCells3D += nz * ystart;
+      if (it.ind == xend)
+        localNumCells3D += nz * ystart;
+      localNumCells3D += nz * ystart;
+    }
+  }
+  return localNumCells3D;
+}
+
+int Mesh::localSize2D() {
+  if (localNumCells2D < 0) {
+    const int xs = firstX() ? xstart - 1 : xstart;
+    const int xe = lastX() ? xend + 2 : xend + 1;
+    const int nx = xe - xs;
+    const int ny = yend - ystart + 1;
+    localNumCells2D = nx * ny;
+    for (RangeIterator it = iterateBndryLowerY(); !it.isDone(); it++) {
+      if (it.ind == xstart)
+        localNumCells2D += ystart;
+      if (it.ind == xend)
+        localNumCells2D += ystart;
+      localNumCells2D += ystart;
+    }
+    for (RangeIterator it = iterateBndryUpperY(); !it.isDone(); it++) {
+      if (it.ind == xstart)
+        localNumCells2D += ystart;
+      if (it.ind == xend)
+        localNumCells2D += ystart;
+      localNumCells2D += ystart;
+    }
+  }
+  return localNumCells2D;
+}
+
+int Mesh::localSizePerp() {
+  if (localNumCellsPerp < 0) {
+    const int xs = firstX() ? xstart - 1 : xstart;
+    const int xe = lastX() ? xend + 2 : xend + 1;
+    const int nx = xe - xs;
+    const int nz = LocalNz;
+    localNumCellsPerp = nx * nz;
+  }
+  return localNumCellsPerp;
+}
+
+int Mesh::globalStartIndex3D() {
+  int localSize = localSize3D();
+  int cumulativeSize;
+  MPI_Scan(&localSize, &cumulativeSize, 1, MPI_INT, MPI_SUM, BoutComm::get());
```

Comment:
I guess these functions are not going to be called a lot? Otherwise, we probably want to cache the value.

Reply:
They will only be called when creating a new indexer object for the mesh. In the case of the global mesh, this will only happen once (at startup). We could cache it for completeness, but in reality it won't be called again anyway.